### ISSUE TITLE
New pathfinder doesn't find paths for all assets in source account (strict-receive).

### ISSUE THREAD

**Issue by marcinx** [stellar/go#1678]
Looks like the new pathfinder doesn't find (existing) paths for all assets trusted in source account.

```
https://horizon.stellar.org/paths/strict-receive?source_account=GASKYWPPQ2VSO6KNIPIRVXMSSDGZLYZQ67CDTLVXOXYDG26SPZ66EDCQ&destination_asset_type=credit_alphanum4&destination_asset_code=USD&destination_asset_issuer=GDUKMGUGDZQK6YHYA5Z6AY2G4XDSZPSZ3SW5UN3ARVMO6QSRDWP5YLEX&destination_amount=15.7
```
-> should return paths for all assets in source account, including trusted LTC and ETH assets, i.e. ETH:GBDEVU63Y6NTHJQQZIKVTC23NWLQVP3WJ2RI2OTSJTNYOIGICST6DUXR and LTC:GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5. However it doesn't return any paths for these assets even though they exist.

Astrograph on the other hand finds the paths just fine and you can compare with below query on https://demo.astrograph.io/graph.

```
{findPaymentPaths(
    sourceAccountID: "GASKYWPPQ2VSO6KNIPIRVXMSSDGZLYZQ67CDTLVXOXYDG26SPZ66EDCQ",
    destinationAsset: "USD-GDUKMGUGDZQK6YHYA5Z6AY2G4XDSZPSZ3SW5UN3ARVMO6QSRDWP5YLEX",
    destinationAmount: "15.7"
) {
sourceAsset {
id
}
destinationAsset {
id
}
sourceAmount
path {
id
}
}
}
```

@tomquisel @tamirms @bartekn

**Comment by marcinx** [comment]
PS: Are the any use cases for strict-send and strict-receive to return paths that are NOT the best path? I can't think of one and at the moment we need to iterate over all detected paths and find the best path manually. I can imagine that everyone who is using paths must do that at the moment.

Astrograph returns only the best path for a given pair. Would it make sense for Horizon to filter the results and only return the best path? This would eliminate the extra step needed in applications utilizing the paths endpoint.

@tomquisel

**Comment by tamirms** [comment]
@marcinx 

It is true that account `GASKYWPPQ2VSO6KNIPIRVXMSSDGZLYZQ67CDTLVXOXYDG26SPZ66EDCQ` has trust lines for `ETH:GBDEVU63Y6NTHJQQZIKVTC23NWLQVP3WJ2RI2OTSJTNYOIGICST6DUXR` and `LTC:GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5`. 

The reason horizon doesn't include any paths originating from `ETH:GBDEVU63Y6NTHJQQZIKVTC23NWLQVP3WJ2RI2OTSJTNYOIGICST6DUXR` or `LTC:GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5` is because the source account doesn't have a sufficient balance to execute any of these payment paths.


Here is the accounts response for `GASKYWPPQ2VSO6KNIPIRVXMSSDGZLYZQ67CDTLVXOXYDG26SPZ66EDCQ` https://gist.github.com/tamirms/f785de4d0c5f4f9a03a749583012ca95

The balance for `LTC:GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5` is 0:
```
    {
      "balance": "0.0000000",
      "limit": "922337203685.4775807",
      "buying_liabilities": "0.0000000",
      "selling_liabilities": "0.0000000",
      "last_modified_ledger": 24024848,
      "is_authorized": true,
      "asset_type": "credit_alphanum4",
      "asset_code": "LTC",
      "asset_issuer": "GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5"
    }
```

There is a postive balance for `ETH:GBDEVU63Y6NTHJQQZIKVTC23NWLQVP3WJ2RI2OTSJTNYOIGICST6DUXR` : 

```
    {
      "balance": "0.0515850",
      "limit": "922337203685.4775807",
      "buying_liabilities": "0.0000000",
      "selling_liabilities": "0.0000000",
      "last_modified_ledger": 24102314,
      "is_authorized": true,
      "asset_type": "credit_alphanum4",
      "asset_code": "ETH",
      "asset_issuer": "GBDEVU63Y6NTHJQQZIKVTC23NWLQVP3WJ2RI2OTSJTNYOIGICST6DUXR"
    }
```

But 0.0515850 ETH is still too small of an amount. According to astrograph's path finding response, https://gist.github.com/tamirms/b38c411b6a075915e8c8fdac2db4f434 , you need at least 0.0938128 ETH to execute the path payment

```
      {
        "sourceAsset": {
          "id": "ETH-GBDEVU63Y6NTHJQQZIKVTC23NWLQVP3WJ2RI2OTSJTNYOIGICST6DUXR"
        },
        "destinationAsset": {
          "id": "USD-GDUKMGUGDZQK6YHYA5Z6AY2G4XDSZPSZ3SW5UN3ARVMO6QSRDWP5YLEX"
        },
        "sourceAmount": "0.0938128",
        "path": [
          {
            "id": "native"
          }
        ]
      }
```

**Comment by tamirms** [comment]
@marcinx 

> PS: Are the any use cases for strict-send and strict-receive to return paths that are NOT the best path? I can't think of one and at the moment we need to iterate over all detected paths and find the best path manually. I can imagine that everyone who is using paths must do that at the moment.

You don't need to iterate over all the paths. The paths are grouped by source asset and then sorted by sourceAmount from cheapest to most expensive

**Comment by marcinx** [comment]
@tamirms 

> The reason is because the source account doesn't have a sufficient balance to execute any of these payment paths.
Hmm, that's probably a regression though? It used to return paths for assets that don't have sufficient funding and this change broke our implementation.

We were using /paths in a two-fold way... 
**a) to get the best path between two assets and send the pathPayment** 
Our platform expects paths for all assets in source account to be returned (as was the case before 0.20.0) and then handles the "sufficient funding/non-sufficient funding"-cases on our end. Now assets are missing and that creates a little bit of a fall out. This is an easy fix but annoying and would be unnecessary if /paths would behave the same as before.

**b) as a live exchange rate mechanism between a target asset and all source assets in an account**
More importantly than a) we were able to give /paths a destination amount prior to 0.20.0 and it would return us the required payment amounts for any asset in source account, regardless of whether the funding for the asset is sufficient or not. This was beautiful because we were able to obtain required payment amounts for any asset configured in source account in a single request. This is actually quite an essential feature in our application.

Use Case: We help online businesses to get paid in crypto (and convert received funds into fiat). 
Before this change we were able to do the following:

* Merchant wants 10 USD for a t-shirt.
* We invoke /paths to obtain a list of required payment amounts for assets pre-configured in a source account, e.g. (BTC, ETH, LTC, XRP, XLM, etc.). Those assets are all supported payment methods on our platform and tied to anchors.
* We could then display back to the customer which amounts he must pay in BTC, ETH, etc. to make the transaction and fulfill the 10 USD requirement by the merchant. 
* The customer could then decide which payment option is suitable for him. **His wallet might very well have sufficient funds, even though the source account we used for /paths has not.**
* We would receive the payment by customer and invoke the pathPayment to convert to the settlement currency chosen by merchant (10 USD). 

This was super fast a super fast and elegant solution for us to obtain live order book information for multiple assets supported on our platform in a single request. Unfortunately with assets being filtered out we cannot do that anymore.

Maybe you could introduce a boolean flag "fundable" for each source asset-path pair instead of simply filtering them out when they don't have enough balance? This would help us with our use case while other users (with other use cases) will be able to filter out assets that don't have sufficient balance easily.

**Comment by marcinx** [comment]
> You don't need to iterate over all the paths. The paths are grouped by source asset and then sorted by sourceAmount from cheapest to most expensive.
Okay! I think it used to be random. This is better.

**Comment by tamirms** [comment]
@marcinx 

what if the /paths endpoint allowed you to provide a list of source assets instead of a source account? It seems like that API would fit your use case better

**Comment by marcinx** [comment]
@tamirms 

Thanks for your follow up. Yeah, that would solve our problem equally well and probably make more sense semantically. We would have to change a few things around in our implementation and add different interfaces for Astro / Horizon paths because they would then behave differently. That should be quite doable though and we'd be happy to have the functionality back.

Would that be easier to implement on your end than adding a flag instead of removing items?

PS: Love the speed on the new /paths and how we haven't run into any 500s yet :)

**Comment by tamirms** [comment]
I'm not sure how we will proceed because there are several ways we could address this issue. 

The other solution I'm considering is similar to your suggestion: We could add a field in the response containing paths which were discarded and the reason why each path was not considered (e.g. insufficient balance or cannot satisfy asset liabilities)

We'll need to spend a bit of time weighing the options.

**Comment by marcinx** [comment]
Okay, great. We'll keep using Astro in the meanwhile.

I just checked and your suggestion would only require minimal changes on our end so that would be perfectly fine as well. 

It could also be a mix of both, flag with "discarded info" along with a new parameter source_assets (which would take a comma separated string `USD:GEWE...123,EUR:GABC...XYZ` or something amongst those lines?).

**Comment by tomquisel** [comment]
Hey there! I'm a fan of the interface @tamirms proposed, where the caller can provide a list of assets. It sounds like that most closely meets your use case, and avoids the need to cleverly create an account with trustlines to the appropriate assets.

@marcinx can you also explain the use case a little more? What happens to the user in the case where they don't have a trustline to the path's source asset? Like, how are they able to make use of that path?

**Comment by marcinx** [comment]
@tomquisel 

> interface

Yeah, that interface makes most sense in the context of strict-send vs. strict-receive semantics. Totally works for us and it is actually much better than having to provide a list of source assets from an account. It turns out Astrograph Pathfinder is broken in certain circumstances and we're currently blocked. Would definitely love to see this materialize sooner than later. 

> use-case

The `/paths` endpoint is actually much more than just a tool to extract lowest-cost paths for path payments for us. It is also a super efficient and lean way to extract best conversion rates between assets as they currently exist on the DEX. 

It's not sufficient to use the ticker or last price to obtain a true conversion rate between assets because the actual conversion rate depends on the depth of the order books. Does that make sense? To convert 1,000 USD to BTC I might have to go deep into the order books and buy up offers at a worse rate than what the ticker or last trade price imply. 

The /paths endpoint provides a true mirror of the conversion rate for any given amount of source asset to destination asset in real time. In this context it shouldn't matter whether there is enough source amount of source asset to cover the payment and this is why above interface is important to us.

> What happens to the user in the case where they don't have a trustline to the path's source asset? Like, how are they able to make use of that path?

In our case we use custodial accounts for our merchants, which are preconfigured with all the trustlines that we support. There are several situations in which we need the payment amount/exchange rate as given by /paths, regardless of whether the source account has enough source asset.

The most prominent one is when a merchant wants to sell an item for asset X and amount Y. Our UI/API gives him a list of payment methods and amounts e.g. XLM, native BTC, native ETH, native LTC, native XRP, which he can then display back to their customer along with (off-chain) deposit addresses (as given by our SEP-6 anchoring partners). The customer then chooses a suitable method and initiates the payment (can be off-chain).

This list of required payment amounts is generated by /paths, which gives us the exact price that must be paid in e.g. BTC for us to be able to convert the deposit into the settlement asset X as desired by the merchant. The user is making the payment in native BTC, which is sent to our SEP-6 anchor and then forwarded to us. We then credit the merchant in his preferred settlement asset, e.g USD.

In above situation we need to extract the "sourceAmount" from /paths for asset BTC:XYZ and it should be irrelevant whether the sourceAccount we pass has enough coverage. The payment is not coming from that given source account but actually from the native Bitcoin network (and then sent to us by anchor and converted by us, etc.).

I hope that makes sense. Hit me up on Keybase if this is still unclear and I will provide more context in DM. The interface as proposed by @tamirms would totally cover our use case as it is independent from any amounts in a source account and the paths are only based on arbitrary assets. Would love to see that (re-)added soon :)

Cheers and thanks in advance!

**Comment by bartekn** [comment]
Closed in #1754. Deployed to staging server: https://horizon-blue.stellar.org/
